modules/040-node-manager/hooks/yandex_preemptibly_delete_preemtible_instances.go

```diff
@@ -0,0 +1,224 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	preemtibleVMDeletionDuration = 24 * time.Hour
+)
+
+type Machine struct {
+	Name              string
+	CreationTimestamp metav1.Time
+	Terminating       bool
+	MachineClassKind  string
+	MachineClassName  string
+}
+
+type YandexInstanceClass struct {
+	Name        string
+	Preemptible bool
+}
+
+func applyMachineFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var terminating bool
+	if obj.GetDeletionTimestamp() != nil {
+		terminating = true
+	}
+
+	classKind, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "class", "kind")
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(classKind) == 0 {
+		return nil, fmt.Errorf("spec.class.kind is empty in %q", obj.GetName())
+	}
+
+	className, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "class", "name")
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(className) == 0 {
+		return nil, fmt.Errorf("spec.class.name is empty in %q", obj.GetName())
+	}
+
+	return &Machine{
+		Name:              obj.GetName(),
+		CreationTimestamp: obj.GetCreationTimestamp(),
+		Terminating:       terminating,
+		MachineClassKind:  classKind,
+		MachineClassName:  className,
+	}, nil
+}
+
+func isPreemptibleFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	preemptible, ok, err := unstructured.NestedBool(obj.UnstructuredContent(), "spec", "preemptible")
+	if err != nil {
+		return nil, fmt.Errorf("can't access field \"preemptible\" of YandexInstanceClass %q: %s", obj.GetName(), err)
+	}
+
+	preemptible = ok && preemptible
+
+	return &YandexInstanceClass{
+		Name:        obj.GetName(),
+		Preemptible: preemptible,
+	}, nil
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Queue:        "/modules/cloud-provider-yandex/preemtibly-delete-preemtible-instances",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "machines",
+			ApiVersion: "machine.sapcloud.io/v1alpha1",
+			Kind:       "Machine",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-cloud-instance-manager"},
+				},
+			},
+			FilterFunc: applyMachineFilter,
+		},
+		{
+			Name:       "ics",
+			ApiVersion: "deckhouse.io/v1",
+			Kind:       "YandexInstanceClass",
```

Comment:
There is a problem because we have no such kind in the cluster if it is not a yandex cloud cluster.

Reply:
I've tested a combination of Kubernetes binding parameters. It'll complain now, but won't stop the main queue. It'll get better once node-manager's CRDs are deployed.

Reply:
Moved everything to the node-manager module and switched to using MachineClass CR.